horovod/common/ops/mpi_gpu_operations.cc

```diff
@@ -148,20 +157,36 @@ Status MPI_GPUAllgather::Execute(std::vector<TensorTableEntry>& entries, const R
   }
 
   timeline.ActivityStartAll(entries, ALLOCATE_OUTPUT);
-  Status status = AllocateOutput(entries, response, entry_component_sizes, recvcounts);
+  Status status = AllocateOutput(entries, response, entry_component_sizes);
   if (!status.ok()) {
+    for (size_t ec = 0; ec < entries.size(); ++ec) {
```

Comment:
Great catch on this memory leak! Unsure why this file is using `new` and `delete` rather than `std::vector` but guess it's always been that way. No need to change it now since the path is very rarely used anyway.

Reply:
I think all Allgather implementations use these manually allocated buffers (and they had leaks originally, #1845). There's really no reason not to use `std::vector` here as the buffers are zero-initialized anyway.

Refactoring these and all helper functions of `AllgatherOp` to use `std::vector` makes a lot of sense, but I'd suggest to do that in an extra PR.

Reply:
Yes, no need to fix it in this PR. I'm not sure if it really needs to be fixed, given that it now works fine (after your memory leak fix).

Reply:
Agree that there's no urgent need to refactor. It would mostly make future changes safer and easier.